checkov/common/graph/graph_builder/variable_rendering/renderer.py

```diff
@@ -0,0 +1,100 @@
+from checkov.common.graph.graph_builder.utils import calculate_hash, run_function_multithreaded
+
+if TYPE_CHECKING:
+    from checkov.common.graph.graph_builder.local_graph import LocalGraph
+
+ATTRIBUTES_NO_EVAL = ["template_body", "template"]
+
+
+class VariableRenderer(ABC):
+    MAX_NUMBER_OF_LOOPS = 50
+
+    def __init__(self, local_graph: "LocalGraph") -> None:
+        self.local_graph = local_graph
+        run_async = os.environ.get("RENDER_VARIABLES_ASYNC", "False")
+        self.run_async = True if run_async == "True" else False
+        self.max_workers = int(os.environ.get("RENDER_ASYNC_MAX_WORKERS", 50))
+        self.done_edges_by_origin_vertex: Dict[int, List[Edge]] = {}
+        self.replace_cache: List[Dict[str, Any]] = [{}] * len(local_graph.vertices)
+
+    def render_variables_from_local_graph(self) -> None:
+        # find vertices with out-degree = 0 and in-degree > 0
+        end_vertices_indexes = self.local_graph.get_vertices_with_degrees_conditions(
+            out_degree_cond=lambda degree: degree == 0, in_degree_cond=lambda degree: degree > 0
+        )
+
+        # all the edges entering `end_vertices`
+        edges_to_render = self.local_graph.get_in_edges(end_vertices_indexes)
+        loops = 0
+        while len(edges_to_render) > 0:
+            logging.info(f"evaluating {len(edges_to_render)} edges")
+            # group edges that have the same origin and label together
+            edges_groups = self.group_edges_by_origin_and_label(edges_to_render)
+            if self.run_async:
+                run_function_multithreaded(
+                    func=self._edge_evaluation_task,
+                    data=edges_groups,
+                    max_group_size=1,
+                    num_of_workers=self.max_workers,
+                )
+            else:
+                for edge_group in edges_groups:
+                    self._edge_evaluation_task([edge_group])
+            for edge in edges_to_render:
+                origin = edge.origin
+                if origin not in self.done_edges_by_origin_vertex:
+                    self.done_edges_by_origin_vertex[origin] = []
+                self.done_edges_by_origin_vertex[origin].append(edge)
+
+            for edge in edges_to_render:
+                origin_vertex_index = edge.origin
+                out_edges = self.local_graph.out_edges.get(origin_vertex_index, [])
+                if all(e in self.done_edges_by_origin_vertex.get(origin_vertex_index, []) for e in out_edges):
+                    end_vertices_indexes.append(origin_vertex_index)
+            edges_to_render = self.local_graph.get_in_edges(end_vertices_indexes)
+            edges_to_render = list(
+                set(
+                    [
+                        edge
+                        for edge in edges_to_render
+                        if edge not in self.done_edges_by_origin_vertex.get(edge.origin, [])
+                    ]
+                )
```

Comment:
```suggestion
                {
                    edge
                    for edge in edges_to_render
                    if edge not in self.done_edges_by_origin_vertex.get(edge.origin, [])
                }
```
give set comprehension some love 😄